File: ProfileDriveTest/src/Subsystems/HIDControllerSubsystem.cpp
/*
 * HIDControllerSubsystem.cpp
 *
 *  Created on: Mar 9, 2017
 *      Author: roberthilton
 */

#include <Subsystems/HIDControllerSubsystem.h>

using namespace frc;
using namespace std;

HIDControllerSubsystem *HIDControllerSubsystem::instance = NULL;
HIDControllerSubsystem::HIDControllerSubsystem() {

	ds = &DriverStation::GetInstance();

	Controllers *robotControllers = Controllers::getInstance();
	driveJoystick = robotControllers->getDriveJoystick();

	shiftAction = new ShiftAction();

	driveBaseSubsystem = DriveBaseSubsystem::getInstance();

	driveHelper = new DriveHelper();

	runThread = false;
	comingFromAuto = true;

	x = 0, y = 0, absLeft = 0, absRight = 0,normalLeft = 0,normalRight = 0, left = 0, right = 0;

	driveJoystickThreadControlStart = 0;
	driveJoystickThreadControlEnd = 0;
	driveJoystickThreadControlElapsedTimeMS = 0;

}

HIDControllerSubsystem* HIDControllerSubsystem::getInstance() {
	if(instance == NULL)
		instance = new HIDControllerSubsystem();

	return instance;
}

HIDControllerSubsystem* HIDControllerSubsystem::getInstance(vector<CustomSubsystem *> *subsystemVector) {
	subsystemVector->push_back(getInstance());
	return instance;
}


void HIDControllerSubsystem::init() {

}

void HIDControllerSubsystem::start() {
	runThread = true;
	driveJoystickThread = thread(&HIDControllerSubsystem::runDriveJoystickThread, this);
}

void HIDControllerSubsystem::subsystemHome() {

}

void HIDControllerSubsystem::stop() {
	runThread = false;
	if(driveJoystickThread.joinable())
		driveJoystickThread.join();
}

void HIDControllerSubsystem::runDriveJoystickThread() {
	while (!ds->IsEnabled()) {this_thread::sleep_for(chrono::milliseconds(20));}
	subsystemHome();

	while (ds->IsAutonomous()) {
		this_thread::sleep_for(chrono::milliseconds(100));
	}

	while(runThread) {
		driveJoystickThreadControlStart = Timer::GetFPGATimestamp();

		if (driveJoystick->GetRisingEdgeButton(DRIVE_SHIFT_LOW)) {
			shiftAction->start(false);
		} else if (driveJoystick->GetRisingEdgeButton(DRIVE_SHIFT_HIGH)) {
			shiftAction->start(true);
		}

		x = driveJoystick->GetRawAxis(DRIVE_X_AXIS);
		y = -driveJoystick->GetRawAxis(DRIVE_Y_AXIS);

		if (!driveBaseSubsystem->isAutoUpdate()) {
			driveBaseSubsystem->setDriveSpeed(driveHelper->calculateOutput(y, x, driveJoystick->GetRawButton(DRIVE_IMM_TURN), driveBaseSubsystem->isHighGear()));
			//driveBaseSubsystem->setDriveSpeed((y + x) * 2300, (y - x) * 2300);
		}

		do {
			driveJoystickThreadControlEnd = Timer::GetFPGATimestamp();
			driveJoystickThreadControlElapsedTimeMS = (int) ((driveJoystickThreadControlEnd - driveJoystickThreadControlStart) * 1000);
			if (driveJoystickThreadControlElapsedTimeMS < MIN_HID_THREAD_LOOP_TIME_MS)
				this_thread::sleep_for(chrono::milliseconds(MIN_HID_THREAD_LOOP_TIME_MS - driveJoystickThreadControlElapsedTimeMS));
		} while(driveJoystickThreadControlElapsedTimeMS < MIN_HID_THREAD_LOOP_TIME_MS);

	}
}

File: RobotJavaProject/src/org/usfirst/frc/team195/robot/Subsystems/HIDControllerSubsystem.java
package org.usfirst.frc.team195.robot.Subsystems;

import java.lang.Thread;

import java.util.List;

import org.usfirst.frc.team195.robot.Actions.IntakePositionAction;
import org.usfirst.frc.team195.robot.Actions.ShiftAction;
import org.usfirst.frc.team195.robot.Reporters.ConsoleReporter;
import org.usfirst.frc.team195.robot.Reporters.MessageLevel;
import org.usfirst.frc.team195.robot.Utilities.*;
import org.usfirst.frc.team195.robot.Utilities.CubeHandler.IntakeControl;

import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj.Timer;

public class HIDControllerSubsystem implements CustomSubsystem {
	
	private static HIDControllerSubsystem instance = null;
	private static final int MIN_HID_THREAD_LOOP_TIME_MS = 20;

	public static HIDControllerSubsystem getInstance() {
		if(instance == null) {
			try {
				instance = new HIDControllerSubsystem();
			} catch (Exception ex) {
				ConsoleReporter.report(ex, MessageLevel.DEFCON1);
			}
		}
		
		return instance;
	}
	
	public static HIDControllerSubsystem getInstance(List<CustomSubsystem> subsystemList) {
		subsystemList.add(getInstance());
		return instance;
	}
	
	@Override
	public void init() {
		;
	}

	@Override
	public void subsystemHome() {
		;
	}

	@Override
	public void start() {
		runThread = true;
		if (!driveJoyStickThread.isAlive())
			driveJoyStickThread.start();
	}

	@Override
	public void terminate() {
		ConsoleReporter.report("CAN'T STOP, WON'T STOP, DON'T CALL ME!", MessageLevel.ERROR);
//		runThread = false;
//		try {
//			driveJoyStickThread.join(Constants.kThreadJoinTimeout);
//		} catch (Exception ex) {
//			ConsoleReporter.report(ex);
//		}
	}
	
//	@Override
//	public String toString() {
//		return generateReport();
//	}
	
	private HIDControllerSubsystem() throws Exception {
		super();
		ds = DriverStation.getInstance();

		Controllers robotControllers = Controllers.getInstance();
		driveJoystick = robotControllers.getDriveJoystick();

		driveBaseSubsystem = DriveBaseSubsystem.getInstance();
		cubeHandlerSubsystem = CubeHandlerSubsystem.getInstance();

		runThread = false;
		comingFromAuto = true;
		
		driveJoyStickThread = new DriveJoyStickThread();
	}
	
	private DriveJoyStickThread driveJoyStickThread;
	
	protected DriveBaseSubsystem driveBaseSubsystem;
	protected CubeHandlerSubsystem cubeHandlerSubsystem;
	protected DriverStation ds;
	protected KnightJoystick driveJoystick;
	
	protected boolean runThread;

	protected boolean comingFromAuto;
	
	private class DriveJoyStickThread extends Thread {
		private ThreadRateControl threadRateControl = new ThreadRateControl();
		
		private DriveHelper driveHelper;
		private ShiftAction shiftAction;
		private IntakePositionAction intakePositionAction;

		private double x, y;
		
		public DriveJoyStickThread() {
			driveHelper = new DriveHelper();
			shiftAction = new ShiftAction();
			intakePositionAction = new IntakePositionAction();
		}
		
		@Override
		public String toString() {
			String retVal = "";
			//retVal += "DriverXAxis:" + driveJoystick.getRawAxis(Constants.DRIVE_X_AXIS) + ";";
			//retVal += "DriverYAxis:" + driveJoystick.getRawAxis(Constants.DRIVE_Y_AXIS) + ";";
			return retVal;
		}
		
		@Override
		public void run() {
			while (!ds.isEnabled()) {try{Thread.sleep(20);}catch(Exception ex) {}}

			while (ds.isAutonomous()||ds.isTest()) {try {Thread.sleep(100);} catch (Exception ex) {}}

			threadRateControl.start();

			while(runThread) {
				if (driveJoystick.GetRisingEdgeButton(Constants.DRIVE_SHIFT_LOW)) {
					shiftAction.start(false);
				} else if (driveJoystick.GetRisingEdgeButton(Constants.DRIVE_SHIFT_HIGH)) {
					shiftAction.start(true);
				}

				/*if (driveJoystick.getRawButton(Constants.INTAKE_CLOSE_RUN)) {
					shiftAction.start(false);
					cubeHandlerSubsystem.setIntakeControl(IntakeControl.FORWARD);
				} else if (driveJoystick.GetRisingEdgeButton(Constants.INTAKE_OPEN)) {
					shiftAction.start(true);
				} else if (driveJoystick.getRawButton(Constants.INTAKE_RUN_REVERSE)) {
					cubeHandlerSubsystem.setIntakeControl(IntakeControl.REVERSE);
				} else if (driveJoystick.getRawButton(Constants.INTAKE_CLOSE)) {
					shiftAction.start(false);
				} else if(driveJoystick.getRawButton(Constants.INTAKE_RUN)) {
					cubeHandlerSubsystem.setIntakeControl(IntakeControl.FORWARD);
				} else {
					cubeHandlerSubsystem.setIntakeControl(IntakeControl.OFF);
				}*/
				
//				if (driveJoystick.getRawButton(Constants.INTAKE_CLOSE)) {
//					shiftAction.start(true);
//					intakePositionAction.start(false);
//				} else if (driveJoystick.getRawButton(Constants.INTAKE_OPEN)) {
//					intakePositionAction.start(true);
//					shiftAction.start(false);
//				} else if (driveJoystick.getRawButton(Constants.INTAKE_CLOSE_HALF)) {
//					intakePositionAction.start(true);
//					shiftAction.start(true);
//				} else if (driveJoystick.getRawButton(Constants.INTAKE_RUN_REVERSE)) {
//					cubeHandlerSubsystem.setIntakeControl(IntakeControl.REVERSE);
//				} else if(driveJoystick.getRawButton(Constants.INTAKE_RUN)) {
//					cubeHandlerSubsystem.setIntakeControl(IntakeControl.FORWARD);
//				} else {
//					cubeHandlerSubsystem.setIntakeControl(IntakeControl.OFF);
//				}

				x = driveJoystick.getRawAxis(Constants.DRIVE_X_AXIS);
				y = -driveJoystick.getRawAxis(Constants.DRIVE_Y_AXIS);

				driveBaseSubsystem.setBrakeMode(driveJoystick.getRawButton(Constants.DRIVE_HOLD_BRAKE));

				driveBaseSubsystem.setDriveOpenLoop(driveHelper.calculateOutput(y, x, driveJoystick.getRawButton(Constants.DRIVE_IMM_TURN), driveBaseSubsystem.isHighGear()));
				//driveBaseSubsystem.setDriveSpeed((y + x) * 2300, (y - x) * 2300);

				threadRateControl.doRateControl(MIN_HID_THREAD_LOOP_TIME_MS);
			}
		}
	}

//	@Override
//	public String generateReport() {
//		String retVal = "";
//		retVal += driveJoyStickThread.toString();
//		return retVal;
//	}
}


